kernel/condvar.c

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <toolchain.h>
+#include <ksched.h>
+#include <wait_q.h>
+
+static struct k_spinlock lock;
+
+int z_impl_k_condvar_init(struct k_condvar *condvar)
+{
+	z_waitq_init(&condvar->wait_q);
+	z_object_init(condvar);
+	return 0;
+}
+
+int z_impl_k_condvar_signal(struct k_condvar *condvar)
+{
+	int woken = 0;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	struct k_thread *thread = z_unpend_first_thread(&condvar->wait_q);
+
+	if (thread != NULL) {
+		arch_thread_return_value_set(thread, 0);
+		z_ready_thread(thread);
+		woken = 1;
+
+	}
+	z_reschedule(&lock, key);
+	return woken;
+}
+
+int z_impl_k_condvar_broadcast(struct k_condvar *condvar)
+{
+	struct k_thread *pending_thread;
+	k_spinlock_key_t key;
+	int woken = 0;
+
+	key = k_spin_lock(&lock);
+
+	/* wake up any threads that are waiting to write */
+	while ((pending_thread = z_unpend_first_thread(&condvar->wait_q)) !=
+	       NULL) {
+		arch_thread_return_value_set(pending_thread, 0);
+		z_ready_thread(pending_thread);
+		woken++;
+	}
+
+	z_reschedule(&lock, key);
+
+	return woken;
```

Comment:
Same issue here, the return value needs to be specified better in the documentation.

Reply:
Yes: this is exactly why "non-negative is success" is so useful: to return the number of things affected.

Reply:
The posix counterparts return only 0 for success and <0 for error. Do we really need to know the count here?

Reply:
It's useful in testing, since there's no other indicator that something did or did not happen.  But if we're striving to match an existing API, maybe it's not right to do that.

Reply:
If we have in the future `pthread_cond_xxx()` API which returns <0 on error and 0 on success, then it should be quite trivial to check that if `k_condvar_signal/broadcast()` returns >= 0, then return 0 in the pthread variant. So the zephyr API does not need to match exactly to pthread one in this case.